### ISSUE TITLE
Espresso sometimes does not wait enough for RecyclerView with setHasFixedSize(true) to load

### ISSUE THREAD

**Issue by jongerrish** [android/android-test#223]
Steps to reproduce:

1. Open attached project
2. Run ExampleInstrumentedTest couple of times, at least one test will fail eventually

If you add Thread.Sleep to the test or remove setHasFixedSize(true), tests will stop failing
[RecyclerTest.zip](https://github.com/android/android-test/files/2927839/RecyclerTest.zip)

Migrated from: https://issuetracker.google.com/123653014

**Comment by mrk-han** [comment]
@jongerrish Reproduced this with your test project, but I noticed it is using the Support Library

Is this reproducible with AndroidX APIs as well?

(BTW, I Enjoyed your Android Dev Summit Testing Rebooted talk a lot 😄 )

**Comment by matejdro** [comment]
Yes, this happens on AndroidX as well. The test project is just most basic reproducible project and Android Studio uses support library by default when creating new project.

**Comment by mrk-han** [comment]
Are the APIs identical? If not, then I think probably best to use a test project using AndroidX, no? Maybe I'm missing something here.

**Comment by matejdro** [comment]
Yes, APIs are identical.

**Comment by tsollas** [comment]
I'm seeing this issue as well in an AndroidX project. I ported @matejdro's example source to AndroidX and can reproduce the problem.

[RecyclerTest-AndroidX.zip](https://github.com/android/android-test/files/3212876/RecyclerTest-AndroidX.zip)

Note that this occurs with 1.1.1 (espresso 3.1.1) as well as the current (as of this writing) 1.2.0-beta-01 (espresso 3.2.0-beta01).

**Comment by matejdro** [comment]
For anyone looking for a workaround, I came up with this:

```kotlin
/**
 * Stop the test until RecyclerView's data gets loaded.
 *
 * Passed [recyclerProvider] will be activated in UI thread, allowing you to retrieve the View.
 *
 * Workaround for https://issuetracker.google.com/issues/123653014
 */
inline fun waitUntilLoaded(crossinline recyclerProvider: () -> RecyclerView) {
    Espresso.onIdle()

    lateinit var recycler: RecyclerView

    InstrumentationRegistry.getInstrumentation().runOnMainSync {
        recycler = recyclerProvider()
    }

    while (recycler.hasPendingAdapterUpdates()) {
        Thread.sleep(10)
    }
}
```

Then just call the method before verifying recycler stuff, passng recycler via lambda: 
```kotlin
waitUntilLoaded { fragment.recycler }
```

**Comment by jonqchk** [comment]
Is there any progress for this issue?

**Comment by sagix** [comment]
Another workaround is to set the itemAnimator to null if animation are disabled:

```kotlin
        val animatorDurationScale = Settings.Global.getFloat(
            context.contentResolver, Settings.Global.ANIMATOR_DURATION_SCALE, 1F
        )
        recyclerView.itemAnimator = if (animatorDurationScale == 0F) {
            null
        } else {
            DefaultItemAnimator()
        }
```

**Comment by MichalDanielDobrzanski** [comment]
Any update on this?

**Comment by vinaygopinath** [comment]
I believe this is still an issue. One option would be to define a [RecyclerView idling resource](https://www.javacodegeeks.com/2019/08/espresso-idling-resource-recyclerview-data-changes.html) but that requires significant changes to the source code just to support testing, something I'd rather not do.

Are folks using the workarounds in this thread?
